src/dart/lib/decoder.dart

```diff
@@ -0,0 +1,894 @@
+    List<Color>? palette,
+    required double height,
+    required this.clipBehavior,
+  }) : assert(palette == null || palette.length <= kCREGLength) {
+    checkSignature();
+    if (palette != null) {
+      applyCustomPalette(palette);
+    }
+    readMetadata();
+    if (foundPalette) {
+      assert(CREG.length == customPalette.length);
+      CREG.setRange(0, customPalette.length, customPalette);
+    }
+    if (!foundViewBox) {
+      minX = -32.0;
+      maxX = 32.0;
+      minY = -32.0;
+      maxY = 32.0;
+    }
+    viewBox = Rect.fromLTRB(minX, minY, maxX, maxY);
+    H = height.isFinite ? height : maxY - minY; // TODO(ianh): consider the conclusions in https://github.com/google/iconvg/issues/20
+    execute();
+  }
+
+  _IconVGMachine._testDecoders(this.bytes, List<Color> palette) : H = 0.0, clipBehavior = Clip.none {
+    applyCustomPalette(palette);
+    assert(foundPalette);
+    assert(CREG.length == customPalette.length);
+    CREG.setRange(0, customPalette.length, customPalette);
+  }
+
+  final Clip clipBehavior;
+  
+  late final Picture picture;
+  late final Rect viewBox;
+  
+  late final double minX, maxX, minY, maxY;
+  bool foundViewBox = false;
+  
+  final Uint8List bytes;
+  int cursor = 0;
+  
+  int get nextByte {
+    if (cursor >= bytes.length) {
+      throw FormatException('Unexpected end of file at offset $cursor.');
+    }
+    return bytes[cursor++];
+  }
+
+  static int ColorToRGBA(Color color) {
+    final int alpha = color.value >> 24;
+    final int red = (((color.value & 0x00FF0000) >> 16) * alpha) ~/ 255;
+    final int green = (((color.value & 0x0000FF00) >> 8) * alpha) ~/ 255;
+    final int blue = ((color.value & 0x000000FF) * alpha) ~/ 255;
+    return (red << 24) + (green << 16) + (blue << 8) + alpha;
+  }
+  
+  static Color RGBAToColor(int color) {
+    final int alpha = color & 0x000000FF;
+    if (alpha == 0x00) {
+      // We sometimes (for gradients) add in real color to the zero-alpha colors.
+      // These don't need to be unmultiplied further, they're already unmultiplied.
+      return Color((color & 0xFFFFFF00) >> 8);
+    }
+    assert(!isNonColor(color));
+    assert(((color & 0xFF000000) >> 24) <= alpha);
+    final int red = (((color & 0xFF000000) >> 24) * 255) ~/ alpha;
+    assert(((color & 0x00FF0000) >> 16) <= alpha);
+    final int green = (((color & 0x00FF0000) >> 16) * 255) ~/ alpha;
+    assert(((color & 0x0000FF00) >> 8) <= alpha);
+    final int blue = (((color & 0x0000FF00) >> 8) * 255) ~/ alpha;
+    return Color((alpha << 24) + (red << 16) + (green << 8) + blue);
+  }
+  
+  static const int kCREGLength = 64;
+  static const int kNREGLength = 64;
+  
+  final Uint32List customPalette = Uint32List.fromList(List<int>.filled(kCREGLength, 0x000000FF));
+  bool foundPalette = false;
+
+  final Uint32List CREG = Uint32List.fromList(List<int>.filled(kCREGLength, 0x000000FF));
+  final Float32List NREG = Float32List(kNREGLength);
+  late final double H;
```

Comment:
Consider using dart style on these (e.g. `creg`, `nreg`, `h`, but spelled out fully without abbreviations).

Reply:
I went with matching the spec for sanity.